qiskit/scheduler/methods/basic.py

```diff
@@ -145,31 +147,46 @@ def translate_gates_to_pulse_defs(circuit: QuantumCircuit,
     circ_pulse_defs = []
 
     cmd_def = schedule_config.cmd_def
-    measured_qubits = set()  # Collect qubits that would like to be measured
+    m_slots = {}  # Map measured qubit index to clbit
 
     def get_measure_schedule() -> CircuitPulseDef:
         """Create a schedule to measure the qubits queued for measuring."""
         measures = set()
         all_qubits = set()
         sched = Schedule()
-        for q in measured_qubits:
-            measures.add(tuple(schedule_config.meas_map[q]))
+        for qubit in m_slots.keys():
+            measures.add(tuple(schedule_config.meas_map[qubit]))
         for qubits in measures:
+            unused_mem_slots = set(qubits) - set(m_slots.values())
             all_qubits.update(qubits)
-            # TODO (Issue #2704): Respect MemorySlots from the input circuit
-            sched |= cmd_def.get('measure', qubits)
-        measured_qubits.clear()
+            default_sched = cmd_def.get('measure', qubits)
+            for time, inst in default_sched.instructions:
+                if isinstance(inst, AcquireInstruction):
+                    mem_slots = []
+                    for channel in inst.acquires:
+                        if channel.index in m_slots.keys():
+                            mem_slots.append(MemorySlot(m_slots[channel.index]))
+                        else:
+                            mem_slots.append(MemorySlot(unused_mem_slots.pop()))
+                    new_acquire = AcquireInstruction(command=inst.command,
+                                                     acquires=inst.acquires,
+                                                     mem_slots=mem_slots)
+                    sched |= new_acquire << time
+                else:
```

Comment:
Let's make this `elif` by combining with the next line. Then, let's add some kind of comment like `#  Measurement pulses should only be added if its qubit was measured by the user`